### PR TITLE
Remove threshold factor from reppoe

### LINKED ISSUE #1960: Review if maximum DR eligibility is working properly
In the commit bd03469652cfb6007917c83f38e66bd201b78510 of the `202106-recovery` branch, a new maximum data request eligibility was included. However, some logs were seen by @tmpolaczyk stating that the probability for mining a data request was 100%.

Some of the tasks that should be covered by this issue:
- [x] Code review of the DR maximum eligibility logic
- [x] Implement TODO at `node/src/actors/chain_manager/mining.rs:325` (pass difficulty as argument)
- [x] Implement some tests that check new DR maximum eligibility logic (after third hard-fork)
- [x] Identify and fix the problem of the mistmatch between the new maximum DR elelibility and the 100% prob. log

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Implement https://github.com/witnet/WIPs/pull/58

This branch is based on the tapi2 branch, however the automated WIP activation is not handled in this PR (`active_wips.wip0016()` will always return false).

Close #1960

### CODE REVIEW

File: validations/src/tests.rs [hunk 1]
```diff
@@ -2326,6 +2326,7 @@ fn test_empty_commit(c_tx: &CommitTransaction) -> Result<(), failure::Error> {
     let collateral_minimum = 1;
     let collateral_age = 1;
     let block_number = 0;
+    let minimum_reppoe_difficulty = 1;
```
Comment (mariocao) [inline validations/src/tests.rs:2329]: Smart to avoid chainging the tests! :smile:

Comment (tmpolaczyk) [reply validations/src/tests.rs:2329]: The idea is to keep old tests working exactly like before and add new tests checking the new formula, but I don't know how to do that in a maintainable way.

File: validations/src/validations.rs [hunk 2]
```diff
@@ -3569,3 +3388,665 @@ mod tests {
...
+        let mut rep_engine = ReputationEngine::new(1000);
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+        rep_engine
+            .trs_mut()
+            .gain(Alpha(10), vec![(id1, Reputation(50))])
+            .unwrap();
+        rep_engine.ars_mut().push_activity(vec![id1]);
+
+        // 0.05% when the total reputation is less than 2000
+        let (t00, p00) = calculate_reppoe_threshold_v2(&rep_engine, &id1, 1, 2000);
```
Comment (mariocao) [inline validations/src/validations.rs:3524]:
```suggestion
        let (t00, _p00) = calculate_reppoe_threshold_v2(&rep_engine, &id1, 1, 2000);
```

File: validations/src/validations.rs [hunk 3]
```diff
@@ -3569,3 +3388,665 @@ mod tests {
...
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFF / 2000));
+        // 1 / 2000 = 500 / 1_000_000
+        assert_eq!((p01 * 1_000_000_f64).round() as i128, 500);
+    }
+
+    // FIXME: Allow for now, wait for https://github.com/rust-lang/rust/issues/67058 to reach stable
+    #[allow(clippy::cast_possible_truncation)]
+    #[test]
+    fn reppoe_overflow_v3() {
+        // Test the behavior of the algorithm when our node has 0 reputation
```
Comment (mariocao) [inline validations/src/validations.rs:4036]:
```suggestion
        // Test the behavior of the algorithm when our node has much reputation
```

File: validations/src/validations.rs [hunk 4]
```diff
@@ -3569,3 +3388,665 @@ mod tests {
...
+            .trs_mut()
+            .gain(Alpha(10), vec![(id0, Reputation(u32::max_value() - 2))])
+            .unwrap();
+
+        // Test big values that result in < 100%
+        let (t01, p01) = calculate_reppoe_threshold_v3(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFE));
+        assert_eq!((p01 * 100_f64).round() as i128, 100);
+    }
+}
```
Comment (mariocao) [inline validations/src/validations.rs:4052]: This test was mind blowing for me because despite of being correct, it represents a network condition that is highly unlikely.

It is important to remember that the minimum difficulty is only activated in case of total active reputation is below 2000:

```
let dr_eligibity = identity_eligibility / max(total_active_reputation, minimum_difficulty) * replication_factor
```

This test shows a situation in which there is only 1 active identity with most of reputation. Therefore, the `identity_eligilibity` (aka "trapezoid") is not coming into play as there is no other active reputed identity to redistribute eligibility. IMHO, this case should be a big problem because it would be likely to happen only after several rollbacks. It will be automatically solved after a data request with a replicacion factor greater than 1.

It would be nice to have some other tests checking a situation with more than 1 active reputed identity.

Comment (tmpolaczyk) [reply validations/src/validations.rs:4052]: > IMHO, this case should be a big problem

I assume you mean "should not"

Actually, this test has 2 identities, but 1 of them has 0 reputation so it is not taken into account in the trapezoid. I will add more tests.

**Comment by tmpolaczyk** [comment]
I want to add one test more for the worst case scenario: a data request with the maximum number of witnesses allowed by the current block size, during the last commit round (to take into account backup witnesses), and ensure that the probability for nodes with no reputation is reasonable.

File: validations/src/validations.rs [hunk 5]
```diff
@@ -3569,3 +3388,1162 @@ mod tests {
...
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+
+        let (_t00, p00) = calculate_reppoe_threshold_v3(
+            &rep_engine,
+            &id1,
+            max_witnesses + max_backup_witnesses,
+            minimum_difficulty,
+        );
+
+        assert_eq!((p00 * 100_f64).round() as i128, 32);
```
Comment (tmpolaczyk) [inline validations/src/validations.rs:4547]: This is a test that checks the worst case reppoe probability in mainnet with an empty ARS, a data request with the maximum number of witnesses allowed by the block weight limits, and taking into account the extra commit rounds.

The result is a data request with 126 witnesses (630 after adding the backup witnesses), where a node with no reputation has a 32% chance to create a valid commitment. This 32% will only be used during one epoch, if the previous commit rounds resulted in fewer commits than expected. Here are the probabilities of the other rounds for comparison:

```
32% for round 4
19% for round 3
13% for round 2
9% for round 1
```

Also, it is not possible to include other data requests while a data request with maximum weight is being resolved, so in order to achieve a "round 4" probability in every epoch, the data request would need have around 30 witnesses, to simultaneously have 4 active requests, resulting in a 7% probability in round 4 and proportionally smaller probabilities in the other rounds.

**Comment by tmpolaczyk** [comment]
Github closed this PR because I merged tapi2 into master

**Comment by tmpolaczyk** [comment]
> Github closed this PR because I merged tapi2 into master

And now it thinks that I deleted the master branch:

Captura de pantalla de 2021-06-22 16-01-26

Don't panic, I didn't delete it, I deleted the tapi2 branch which was the old base branch.

File: validations/src/validations.rs [hunk 6]
```diff
@@ -3569,3 +3388,1162 @@ mod tests {
         assert_eq!(a.slot(&h3), 2);
     }
 }
+
+#[cfg(test)]
+mod reppoe_tests {
+    use super::*;
+    use witnet_data_structures::chain::{Alpha, Environment};
```
Comment (lrubiorod) [inline validations/src/validations.rs:3395]: fmt

File: validations/src/validations.rs [hunk 7]
```diff
@@ -3569,3 +3388,1162 @@ mod tests {
...
+            .unwrap();
+    }
+
+    // FIXME: Allow for now, wait for https://github.com/rust-lang/rust/issues/67058 to reach stable
+    #[allow(clippy::cast_possible_truncation)]
+    #[test]
+    fn target_reppoe_v1() {
+        let mut rep_engine = ReputationEngine::new(1000);
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+        rep_engine
```
Comment (lrubiorod) [inline validations/src/validations.rs:3489]: why you dont use the auxiliary add_rep function that you create before

File: validations/src/validations.rs [hunk 8]
```diff
@@ -3569,3 +3388,1162 @@ mod tests {
...
+            .gain(Alpha(10), vec![(id0, Reputation(u32::max_value() - 2))])
+            .unwrap();
+
+        // Test big values that result in < 100%
+        // Active identity with 100% of the reputation
+        let (t01, p01) = calculate_reppoe_threshold_v1(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFE));
+        assert_eq!((p01 * 100_f64).round() as i128, 100);
+        // Active identity with 0 reputation
+        let (t02, p02) = calculate_reppoe_threshold_v3(&rep_engine, &id1, 1, 2000);
```
Comment (lrubiorod) [inline validations/src/validations.rs:4039]: should it be v1?

File: validations/src/validations.rs [hunk 9]
```diff
@@ -3569,3 +3388,1162 @@ mod tests {
...
+            .unwrap();
+
+        // Test big values that result in < 100%
+        // Active identity with 100% of the reputation
+        let (t01, p01) = calculate_reppoe_threshold_v2(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFF / 2000));
+        // 1 / 2000 = 500 / 1_000_000
+        assert_eq!((p01 * 1_000_000_f64).round() as i128, 500);
+        // Active identity with 0 reputation
+        let (t02, p02) = calculate_reppoe_threshold_v3(&rep_engine, &id1, 1, 2000);
```
Comment (lrubiorod) [inline validations/src/validations.rs:4085]: and v2 here?

File: validations/src/validations.rs (hunk 8 above)
Comment (tmpolaczyk) [reply validations/src/validations.rs:4039]: :scream: 

Yes...

**Review by lrubiorod — APPROVED** [review]
